LSM: fix update_tables() logging non-updated tables

kprotty opened the pull request:
@sentientwaffle Discovered that the tables passed into `Manifest.update_tables` were being used to update the `ManifestLevel` table snapshots, but were being inserted into the `ManifestLog` without those updates. This is a proposed fix for that.

Commit:
LSM: fix update_tables() logging non-updated tables

`Manifest.update_tables()` would call set_snapshot_max() with the tables. This would find its own internal tables from those passed in and update their snapshot_max. update_tables() would incorrectly insert to the manifest_log the original tables passed in, not the internal ones with updated snapshot_max. This fixes that.

sentientwaffle commented:
Alternatively, what about:

- `ManifestLevel.set_snapshot_max` took a `[]TableInfo` instead of `[]const TableInfo`. In addition to updating its internal `TableInfo`s' `snapshot_max`, it would update the parameters' `snapshot_max`.
- And `update_tables` also takes `[]TableInfo` instead of `[]const TableInfo`. But since `ManifestLevel.set_snapshot_max` would update the `snapshot_max`, it wouldn't have to be done here.

I like this approach because it keeps the `snapshot_max` updates synced up in one place.

kprotty commented:
This could work as the only thing that calls `update_tables` provides it from a `TableInfoBufferType` which is mutable. I'm not sure of the implications for writing unit tests against Manifest, but I think mutable TableInfos is a fair compromise.

sentientwaffle commented:
Yeah, it does make the unit tests a little trickier since they have to verify that the parameter is mutated correctly. I think it is worthwhile though; otherwise we are relying on `set_snapshot_max` and `update_tables` performing identical mutations in different places one after the other.

(Also, be sure to document on both functions that the parameter is mutated, since that might be unexpected.)

jorangreef commented:
Great suggestion @sentientwaffle — this bug got me thinking that we didn't have the interface right.

Commit:
LSM: update_tables/set_snapshot_max table mutable TableInfos instead

To ensure that the table snapshots are properly updated for all call-sites, update_tables() takes a mutable slice of TableInfos.

kprotty commented:
I rebased on main, could you review again @sentientwaffle ?
